stdlib/domain.ml

```diff
@@ -124,30 +124,45 @@ module DLS = struct
   external set_dls_list : entry list  -> unit
     = "caml_domain_dls_set" [@@noalloc]
 
-  let new_key () = ref 0
+  let new_key f =
+    let k = ref 0 in
+    let vals = get_dls_list () in
+    let e = {key = k; slot = ref (Obj.repr (f ())) } in
+    set_dls_list (e::vals);
+    (k, f)
 
   let set k x =
     let cs = Obj.repr x in
-    let old = get_dls_list () in
+    let vals = get_dls_list () in
     let rec add_or_update_entry k v l =
       match l with
       | [] -> Some {key = k; slot = ref v}
-      | hd::tl -> if (hd.key == k) then begin
-        hd.slot := v;
-        None
-      end else add_or_update_entry k v tl
+      | hd::tl ->
+        if (hd.key == k) then begin
+          hd.slot := v;
+          None
+        end
+        else add_or_update_entry k v tl
     in
-    match add_or_update_entry k cs old with
+    let k = fst k in
```

Comment:
I found this function confusing to follow, because the variable `k` gets rebound from being an `int ref * (unit -> 'a)` to a `int ref`. If you read the function from top to bottom, the `k` coming into the function is not the `k` that is used in `add_or_update_entry`. Can we rename to match what is happening in `get`?

Reply:
done